collect_app/src/main/java/org/odk/collect/android/activities/FormMapActivity.java

```diff
@@ -56,15 +64,45 @@
 
     private FormMapViewModel viewModel;
 
-    private MapFragment map;
+    @VisibleForTesting public MapFragment map;
+
+    /**
+     * Quick lookup of instance objects from map feature IDs.
+     */
+    private final Map<Integer, FormMapViewModel.MappableFormInstance> instancesByFeatureId = new HashMap<>();
+
+    /**
+     * Points to be mapped. Note: kept separately from {@link #instancesByFeatureId} so we can
+     * quickly zoom to bounding box.
+     */
+    private final List<MapPoint> points = new ArrayList<>();
+
+    /**
+     * True if the map viewport has been initialized, false otherwise.
+     */
+    private boolean viewportInitialized;
+
+    @VisibleForTesting public ViewModelProvider.Factory viewModelFactory;
 
     @Override public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
         FormsDao dao = new FormsDao();
-        Form form = dao.getFormsFromCursor(dao.getFormsCursor(getIntent().getData())).get(0);
-        viewModel = ViewModelProviders.of(this,
-                new FormMapActivity.FormMapViewModelFactory(form, new DatabaseInstancesRepository())).get(FormMapViewModel.class);
+        Form form = null;
```

Comment:
@grzesiek2010 this is incidentally a fix for #3546 and the one I would recommend if you feel strongly about addressing it before v1.26 goes out with this change.